### ISSUE TITLE
x/mobile/cmd/gomobile: ordering of manifest attributes may produce invalid manifest

### ISSUE THREAD

**Issue by dskinner** [golang/go#13036]
This subtle difference is not picked up in the tests since `aapt` will still successfully match the parsed xmltree against its own encoding of the same xml. The exact reason for the fault is yet unknown but can be worked around be reordering attributes in an apps manifest.

**Comment by dskinner** [comment]
My initial reaction is to disable sorting of test results. Repeated encoding of the same manifest with the `aapt` tool appears deterministic, so sorting results may be obscuring a more fundamental issue.

```
go test -run BinaryXML -dump
unzip junk.sdk.apk
mv AndroidManifest.xml a.xml
go test -run BinaryXML -dump
unzip junk.sdk.apk
mv AndroidManifest.xml b.xml
diff a.xml b.xml
# no output
```

**Comment by hyangah** [comment]
cc @crawshaw

**Comment by dskinner** [comment]
I've started working to resolve. Will also provide real testing down to byte-to-byte level, testing of any number of manifests, and have it setup so results can be tested against any future version of the sdk build-tools.

**Comment by dskinner** [comment]
I forgot to mention localized testing down to the chunk header as well so any future errors can be more easily identified.

**Comment by dskinner** [comment]
Here's a more formal proposal for how I plan to fix this (most of this is already working, the following being a header in the source):

```
// Binary resource structs support unmarshalling the binary output of aapt.
// Implementations of marshalling for each struct must produce the exact input
// sent to unmarshalling. This allows tests to validate each struct representation
// of the binary format as follows:
//
//  * unmarshal the output of aapt
//  * marshal the struct representation
//  * perform byte-to-byte comparison with aapt output per chunk header and body
//
// This process should strive to make structs idiomatic to make parsing xml text
// into structs trivial.
//
// Once the struct representation is validated, tests for parsing xml text
// into structs can become self-referential as the following holds true:
//
//  * the unmarshalled input of aapt output is the only valid target
//  * the unmarshalled input of xml text may be compared to the unmarshalled
//    input of aapt output to identify errors, e.g. text-trims, wrong flags, etc
//
// This provides validation, byte-for-byte, for producing binary xml resources.
//
// It should be made clear that unmarshalling binary resources is currently only
// in scope for proving that the BinaryMarshaler works correctly. Any other use
// is currently out of scope.
```

**Comment by dskinner** [comment]
this has been addressed with #13109
